libsolidity/analysis/TypeChecker.cpp

```diff
@@ -2549,61 +2571,66 @@ bool TypeChecker::visit(MemberAccess const& _memberAccess)
 					" outside of storage."
 				);
 		}
-		string errorMsg = "Member \"" + memberName + "\" not found or not visible "
-				"after argument-dependent lookup in " + exprType->toString() + ".";
 
-		if (auto const& funType = dynamic_cast<FunctionType const*>(exprType))
-		{
-			auto const& t = funType->returnParameterTypes();
+		auto [errorId, description] = [&]() -> tuple<ErrorId, string> {
+			string errorMsg = "Member \"" + memberName + "\" not found or not visible "
+				"after argument-dependent lookup in " + exprType->toString() + ".";
 
-			if (memberName == "value")
+			if (auto const& funType = dynamic_cast<FunctionType const*>(exprType))
 			{
-				if (funType->kind() == FunctionType::Kind::Creation)
-					errorMsg = "Constructor for " + t.front()->toString() + " must be payable for member \"value\" to be available.";
+				auto const& t = funType->returnParameterTypes();
+
+				if (memberName == "value")
+				{
+					if (funType->kind() == FunctionType::Kind::Creation)
+						return { 8827_error, "Constructor for " + t.front()->toString() + " must be payable for member \"value\" to be available." };
+					else if (
+						funType->kind() == FunctionType::Kind::DelegateCall ||
+						funType->kind() == FunctionType::Kind::BareDelegateCall
+					)
+						return { 8477_error, "Member \"value\" is not allowed in delegated calls due to \"msg.value\" persisting." };
+					else
+						return { 8820_error, "Member \"value\" is only available for payable functions." };
+				}
 				else if (
-					funType->kind() == FunctionType::Kind::DelegateCall ||
-					funType->kind() == FunctionType::Kind::BareDelegateCall
+					t.size() == 1 &&
+					(t.front()->category() == Type::Category::Struct ||
+					t.front()->category() == Type::Category::Contract)
```

Comment:
Not entirely sure what is correct for our coding style. One you wrote or:
```suggestion
					t.size() == 1 && (
						t.front()->category() == Type::Category::Struct ||
						t.front()->category() == Type::Category::Contract
					)
```
or
```suggestion
					t.size() == 1 &&
					(
						t.front()->category() == Type::Category::Struct ||
						t.front()->category() == Type::Category::Contract
					)
```

Reply:
I just kept old formatting. I believe it should be (according to the code style):

```
t.size() == 1 && (
    t.front()->category() == Type::Category::Struct ||
    t.front()->category() == Type::Category::Contract
)
```